src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -2191,6 +2192,21 @@ public ArrayType makeArrayType(Type t) {
     // </editor-fold>
 
     // <editor-fold defaultstate="collapsed" desc="asSuper">
+    /**
+     * Return the (most specific) base type of `t' that starts with the
+     * given type `s'.  If none exists, return null.
+     *
+     * @param t a type
+     * @param s a type
+     */
+    public Type asSuper(Type t, Type s) {
```

Comment:
A possible approach here (discussed offline) would be to leave asSuper as it is today - e.g. assume that asSuper speaks about "extends/implements" clause, and deal with .ref vs. .val mismatches at the callsite (if possible) - e.g. in Types::isSubtypes.

Reply:
Yes, I'll pursue what you outline in upcoming commits